sys/include/riot_hdr.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_riot_hdr RIOT header helpers and tools
+ * @ingroup     sys
+ * @{
+ *
+ * # RIOT image header for riotboot bootloader
+ *
+ * ## Overview
+ *
+ * riotboot is the name of a minimal bootloader application and infrastructure.
+ * It consists of
+ *
+ * - the application "riotboot" in the bootloaders folder which serves as
+ *   minimal bootloader
+ *
+ * - the module "riot_hdr" used to recognise RIOT firmware images
+ *   which riotboot can boot.
+ *
+ * - the module "slot_util" used to manage the images (slots) with a
+ *   RIOT header attached to them
+ *
+ * - a tool in dist/tools/riot_hdr for header generation.
+ *
+ * - several make targets to glue everything together
+ *
+ * ## Concept
+ *
+ * `riotboot` expects the flash to be separated in slots: at CPU_FLASH_BASE
+ * address resides the bootloader, which is followed by a slot 0 with a
+ * firmware image.
+ *
+ * A RIOT image with a single slot looks like:
+ *
+ * |------------------------------- FLASH -------------------------------------|
+ * |----- RIOTBOOT_LEN ----|----------- RIOTBOOT_SLOT_SIZE (slot 0) -----------|
+ *                         |--- RIOTBOOT_HDR_LEN ----|
+ *  ---------------------------------------------------------------------------
+ * |       bootloader      | riot_hdr_t + filler (0) |     RIOT firmware       |
+ *  ---------------------------------------------------------------------------
+ *
+ * Please note that `RIOTBOOT_HDR_LEN` depends on the architecture of the
+ * MCU, since it needs to be aligned to 256B. This is fixed regardless of
+ * `sizeof(riot_hdr_t)`
+ *
+ * The bootloader will, on reset, verify the checksum of the first slot, then
+ * boot it. If the slot doesn't have a valid checksum, no image will be
+ * booted and the bootloader will enter "while(1);".
+ *
+ * The header contains
+ *
+ * - "RIOT" as magic number
+ * - an application version
+ * - a checksum
+ *
+ * The bootloader "riotboot" only cares about checksum.
+ *
+ * @file
+ * @brief       RIOT images header and tools
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Francisco Acosta <francisco.acosta@inria.fr>
+ *
+ * @}
+ */
+
+#ifndef RIOT_HDR_H
+#define RIOT_HDR_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+
+/**
+ *  @brief Number of header bytes that get checksummed
+ */
+#define RIOT_HDR_CHECKSUM_LEN      (12)
+
+/**
+ * @brief Structure to store image header - All members are little endian
+ * @{
+ */
+typedef struct {
```

Comment:
What storage type endianness ?

Reply:
The same as its members.

Reply:
What was the design decision by always having these members as little endian instead of host endian?

Reply:
The endianness should be deterministic because the board reads it but it is generated by your computer.

Reply:
I agree that the endianess should be deterministic.

Either we have to generate a struct on the host with the right (device) endianess or add little endian reads to the bootloader.

Reply:
I asked to use `le32toh` when reading it but @kYc0o said it would make things more complicated… I found it more complicated without.

Reply:
I don't really mind either solution, but IMHO it is not something we can simply ignore.

Reply:
I was in favor of splitting `riot_hdr` in its own PR with its own test to be able to correctly review these things. I asked for it IRL but yeah.